### PR TITLE
Drop return value from Node entrypoint

### PR DESCRIPTION
**Pull request by daviddrysdale**
### Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [x] I have updated [documentation](docs/) accordingly.
  - [x] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by daviddrysdale** [comment]
Fixes #493

File: examples/abitest/module_1/rust/src/lib.rs [hunk 1]
```diff
@@ -25,18 +24,26 @@ use protobuf::ProtobufEnum;
...
-    .value()
+    });
 }
 
-pub fn main(in_handle: u64) -> Result<(), oak::OakStatus> {
+pub fn main(in_handle: u64) {
+    match inner_main(in_handle) {
+        Err(oak::OakStatus::ERR_TERMINATED) => {
+            info!("node terminated");
+        }
```
Comment (tiziano88) [inline examples/abitest/module_1/rust/src/lib.rs]: Personally I think we should get rid of `ERR_TERMINATED` entirely at some point, and reserve `OakStatus` as the return type of ABI function calls. Then I would change the type of `inner_main` to `-> Result<(), Box<dyn Debug>>` or similar, so that arbitrary errors may be handled with `?` throughout the `main` body, and the resulting error may be logged.

This is also similar to how it is allowed to have a [`Result` return value](https://github.com/rust-lang/rust/blob/master/RELEASES.md#language-13) from `main` in standard Rust.

WDYT?

File: oak/server/rust/oak_runtime/src/lib.rs [hunk 2]
```diff
@@ -568,18 +568,29 @@ impl OakNode {
 }
 
 /// Expected type for the main entrypoint to a Node under test.
-pub type NodeMain = fn(Handle) -> Result<(), oak::OakStatus>;
+pub type NodeMain = fn(Handle);
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/lib.rs]: I was wondering whether this could / should move to the `oak_abi` crate at some point?

File: examples/abitest/module_1/rust/src/lib.rs (hunk 1 above)
Comment (daviddrysdale) [reply examples/abitest/module_1/rust/src/lib.rs]: `ERR_TERMINATED` *is* a return type from ABI function calls, particularly `wait_on_channels` &ndash; it indicates that the Node should terminate.  Being propagated out as a return value from `oak_main` was mostly a side effect.

File: oak/server/rust/oak_runtime/src/lib.rs (hunk 2 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/lib.rs]: Yeah, I'd thought that too.  Added an extra commit.

File: examples/abitest/module_0/rust/src/lib.rs [hunk 3]
```diff
@@ -38,22 +38,18 @@ struct FrontendNode {
...
 
 #[no_mangle]
-pub extern "C" fn frontend_oak_main(handle: u64) -> i32 {
-    std::panic::catch_unwind(|| main(handle))
-        .unwrap_or(Err(oak::OakStatus::ERR_INTERNAL))
-        .err()
-        .unwrap_or(oak::OakStatus::OK)
-        .value()
+pub extern "C" fn frontend_oak_main(handle: u64) {
+    let _ = std::panic::catch_unwind(|| main(handle));
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply examples/abitest/module_0/rust/src/lib.rs]: Added `set_panic_hook()`, `oak_log::init` is already called below.

File: examples/chat/module_1/rust/src/lib.rs [hunk 4]
```diff
@@ -17,24 +17,18 @@
...
         // allows it. In this case, the return value of the entry point function will be
         // `ERR_INTERNAL` regardless of the actual error.
-        room.event_loop(handle).expect("terminating with error");
-        Ok(())
-    })
-    .unwrap_or(Err(oak::OakStatus::ERR_INTERNAL))
-    .err()
-    .unwrap_or(oak::OakStatus::OK)
-    .value()
+        let _ = room.event_loop(handle);
```
Comment (tiziano88) [inline examples/chat/module_1/rust/src/lib.rs]: I think the comment does not apply any more here. Also why was the `expect` removed? I think it's fine to leave it here.

File: sdk/rust/oak_derive/src/lib.rs [hunk 5]
```diff
@@ -59,24 +59,20 @@ pub fn derive_oak_exports(input: TokenStream) -> TokenStream {
...
-            .value()
+            });
         }
         // Internal version of the main entrypoint, to allow testing without any
         // panic interception.
-        pub fn inner_main(handle: u64) -> Result<(), oak::OakStatus> {
+        pub fn inner_main(handle: u64) {
             let mut node = <#name>::new();
-            oak::grpc::event_loop(node, oak::ReadHandle{ handle: oak::Handle::from_raw(handle) })
+            let _ = oak::grpc::event_loop(node, oak::ReadHandle{ handle: oak::Handle::from_raw(handle) });
```
Comment (tiziano88) [inline sdk/rust/oak_derive/src/lib.rs]: This should also `unwrap` / `expect` the `Result` here so it gets logged, or it will be lost?

Comment (daviddrysdale) [reply sdk/rust/oak_derive/src/lib.rs]: As with the chat example, the `grpc::event_loop` will only ever return an `Err`, so `expect`ing anything means that Node termination always implies a `panic`.   However, I'm not keen on adding a log statement here because it induces a dependency on the `log` crate for anything that uses `oak_derive`.

Comment (tiziano88) [reply sdk/rust/oak_derive/src/lib.rs]: Couldn't the node terminate gracefully with `Ok` if it wanted to? e.g. in response to a specific gRPC message, or after some time, or for whatever other reason.

But even in the `event_loop`, we do sometimes propagate errors up, e.g.

https://github.com/project-oak/oak/blob/a0ab735c61fa69474b53878e8b9bf406c00a537f/sdk/rust/oak/src/grpc/mod.rs#L174

IIUC, after this commit, this error would just be ignored and never logged. Or am I missing something?

Comment (daviddrysdale) [reply sdk/rust/oak_derive/src/lib.rs]: `grpc::event_loop` does propagate errors up, but that's the only way that loop terminates (atm) &ndash; there's no path to emit `Ok`.  So any Node that uses `oak_derive` will always terminate its inner entrypoint with an `Err`, which would always trigger a `panic` if the `expect` is there, which will (I think) always generate a Wasm trap &ndash; even for orderly shutdown.

Comment (daviddrysdale) [reply sdk/rust/oak_derive/src/lib.rs]: (From discussion) We already transiently depend on the `log` crate anyway (`oak_derive` invokes `oak::set_panic_hook` which calls `log::error!`), so I've added a log here too.

File: examples/abitest/module_0/rust/src/lib.rs [hunk 6]
```diff
@@ -38,22 +38,21 @@ struct FrontendNode {
...
+        oak::set_panic_hook();
+        main(handle)
+    });
 }
 
-pub fn main(handle: u64) -> Result<(), oak::OakStatus> {
+pub fn main(handle: u64) {
     let node = FrontendNode::new();
-    oak::grpc::event_loop(
+    let _ = oak::grpc::event_loop(
```
Comment (tiziano88) [inline examples/abitest/module_0/rust/src/lib.rs]: Handle this result value too (by logging or unwrapping).

File: sdk/rust/oak_derive/src/lib.rs [hunk 7]
```diff
@@ -59,24 +58,22 @@ pub fn derive_oak_exports(input: TokenStream) -> TokenStream {
...
         }
         // Internal version of the main entrypoint, to allow testing without any
         // panic interception.
-        pub fn inner_main(handle: u64) -> Result<(), oak::OakStatus> {
+        pub fn inner_main(handle: u64) {
             let mut node = <#name>::new();
-            oak::grpc::event_loop(node, oak::ReadHandle{ handle: oak::Handle::from_raw(handle) })
+            if let Err(s) = oak::grpc::event_loop(node, oak::ReadHandle{ handle: oak::Handle::from_raw(handle) }) {
+                warn!("Node terminating with {:?}", s);
+            }
```
Comment (tiziano88) [inline sdk/rust/oak_derive/src/lib.rs]: I'm still not convinced this is easier to read or maintain than leaving it as `pub fn inner_main(handle: u64) -> Result<(), oak::OakStatus>` (or in fact change it to `Result<(), Box<dyn Debug>>`, since all we do with the error is log it, and this seems more semantically accurate), and just unwrapping the error on line 67, and letting the panic hook deal with it.

File: examples/abitest/module_0/rust/src/lib.rs (hunk 6 above)
Comment (daviddrysdale) [reply examples/abitest/module_0/rust/src/lib.rs]: Done.

File: sdk/rust/oak_derive/src/lib.rs (hunk 7 above)
Comment (daviddrysdale) [reply sdk/rust/oak_derive/src/lib.rs]: My aim here was to have the `inner_main` is as close to the "real" entrypoint as possible (the only differences being no FFI-required `catch_unwind`, no `extern "C"`).  As [previously mentioned](https://github.com/project-oak/oak/pull/475#discussion_r367936178), this removes the need for the (test) runtime to deal with Rust types (rather than pure Wasm types), which would be needed for a fully Rust-based runtime later.

To put it another way: `NodeMain` moving into `oak_abi` means that it has to have a Wasm-compatible signature, and splitting out `inner_main` was specifically done to expose a `NodeMain` entrypoint for testing.

Comment (tiziano88) [reply sdk/rust/oak_derive/src/lib.rs]: But this still leaves a gap between the two, in that the inner function may panic, while the outer function may not (because of the `catch_unwind`). So you still need to deal with this difference. Anyways, it's fine as it is for now, but I think it would be nice to have clear requirements for the inner and outer functions at some point (e.g. no panics crossing outer function boundary, etc.), and come up with an ergonomic API that satisfies it.
